replicator/src/main/scala/com/evolutiongaming/kafka/journal/replicator/Batch.scala

```diff
@@ -14,139 +24,87 @@ private[journal] object Batch {
-                  delete :: bs
-                }
-
-              case (_: Appends, a: Action.Purge) =>
-                purgeOf(a.origin, a.version) :: Nil
-
-              case (b: Delete, a: Action.Append) =>
-                appendsOf(NonEmptyList.of(actionRecord(a))) :: b :: tail
-
-              case (b: Delete, _: Action.Mark) =>
-                b.copy(offset = offset) :: tail
-
-              case (b: Delete, a: Action.Delete) =>
-                if (a.to > b.to) {
-                  val cleanTail: List[Batch] = tail.mapFilter {
-                    case appends: Appends => dropDeleted(appends, a)
-                    case delete: Delete   => delete.some
-                    case purge: Purge     => purge.some
-                  }
-
-                  val delete = deleteOf(a.to, b.origin orElse a.origin, b.version orElse a.version)
-                  delete :: cleanTail
-                } else {
-                  val delete = b.copy(offset = offset, origin = b.origin orElse a.origin, version = b.version orElse a.version)
-                  delete :: tail
-                }
-
-              case (_: Delete, a: Action.Purge) =>
-                purgeOf(a.origin, a.version) :: Nil
-
-              case (b: Purge, a: Action.Append) =>
-                appendsOf(NonEmptyList.of(actionRecord(a))) :: b :: Nil
-
-              case (b: Purge, _: Action.Mark) =>
-                b.copy(offset = offset) :: Nil
-
-              case (b: Purge, a: Action.Delete) =>
-                deleteOf(a.to, a.origin, a.version) :: b :: Nil
-
-              case (_: Purge, a: Action.Purge) =>
-                purgeOf(a.origin, a.version) :: Nil
+  private final case class State(
+    purge: Option[Purge]     = None,
+    appends: Option[Appends] = None,
+    delete: Option[Delete]   = None,
+  ) {
+
+    def batches: List[Batch] = {
+      // we can drop first `append`, if `deleteTo` will discard it AND there is at least one more `append`
+      val appends = {
+        this.appends match {
+          case Some(appends) =>
+            val deleteTo = this.delete.map(_.to.value)
+            val records  = appends.records
+            val actions =
+              if (deleteTo.contains(records.head.action.range.to) && records.tail.nonEmpty) appends.records.tail
+              else appends.records.toList
+            NonEmptyList.fromList(actions) match {
+              case Some(actions) => appends.copy(records = actions).some
+              case None          => appends.some // cannot happen
             }
 
-          case Nil =>
-            record.action match {
-              case a: Action.Append => appendsOf(NonEmptyList.of(actionRecord(a))) :: Nil
-              case _: Action.Mark   => Nil
-              case a: Action.Delete => deleteOf(a.to, a.origin, a.version) :: Nil
-              case a: Action.Purge  => purgeOf(a.origin, a.version) :: Nil
-            }
+          case None =>
+            None
```

Comment:
This usually means that the underlying `match...case` could be replaced by a simple `flatMap`.